index.js

```diff
@@ -46,184 +45,271 @@ const version = () => {
+          files(file)
+            .then((result) => watcher.add(dependencies(result)))
+            .catch(error)
         })
-        .catch(error)
-      })
     }
   })
   .catch(error)
 
-function processCSS (css, filename) {
-  var spinner = ora(`Processing ${filename || 'your CSS'}`)
-  spinner.start()
-
-  let options = Object.assign(
-    {
-      from: filename,
-      to: output || (argv.replace ? filename : path.join(dir, path.basename(filename)))
-    },
-    config.options
-  )
-  options.to = path.resolve(options.to)
-
-  return postcss(config.plugins).process(css, options)
-    .then(result => {
-      if (result.messages.some(msg => msg.type === 'warning')) spinner.fail()
-
-      var tasks = [fs.outputFile(options.to, result.css)]
-      if (result.map) {
-        tasks.push(fs.outputFile(options.to.replace('.css', '.css.map'), result.map))
-      }
-      return Promise.all(tasks)
-        .then(() => {
-          spinner.succeed()
-          return result
-        })
+function rc (ctx, path) {
+  if (argv.use) return Promise.resolve()
+
+  return postcssrc(ctx, path)
+    .then((rc) => { config = rc })
+    .catch((err) => {
+      if (err.message.indexOf('No PostCSS Config found') === -1) throw err
     })
 }
 
-function processFiles (files) {
-  if (typeof files === 'string') files = [files]
-  return Promise.all(files.map(file => {
-    // Read from stdin
-    if (file === '-') {
-      return getStdin()
-      .then(css => processCSS(css))
+function files (files) {
+  if (typeof files === 'string') files = [ files ]
+
+  return Promise.all(files.map((file) => {
+    if (file === 'stdin') {
+      return stdin()
+        .then((content) => css(content, 'stdin'))
     }
+
     return fs.readFile(file)
-    .then(css => processCSS(css, file))
+      .then((content) => css(content, file))
   }))
 }
 
-function getConfig (ctx, path) {
-  if (argv.use || argv.parser || argv.stringifier || argv.syntax) {
-    config = {
-      plugins: argv.use ? argv.use.map(plugin => require(plugin)) : [],
-      options: {
-        parser: argv.parser ? require(argv.parser) : undefined,
-        syntax: argv.syntax ? require(argv.syntax) : undefined,
-        stringifier: argv.stringifier ? require(argv.stringifier) : undefined
-      }
+function css (css, file) {
+  const ctx = { options: config.options }
+
+  if (file !== 'stdin') {
+    ctx.file = {
+      dirname: path.dirname(file),
+      basename: path.basename(file),
+      extname: path.extname(file)
     }
-    setMap()
-    return Promise.resolve()
-  } else {
-    return postcssrc(ctx, path)
-    .then(conf => {
-      config = conf
-      setMap()
-    })
-    .catch(err => {
-      if (err.message.indexOf('No PostCSS Config found') === -1) throw err
-    })
+
+    if (!argv.config) argv.config = path.dirname(file)
   }
+
+  if (!argv.config) argv.config = process.cwd()
+
+  const time = process.hrtime()
+
+  spinner.text = `Processing ${file}`
+
+  rc(ctx, argv.config)
```

Comment:
:+1: Push code as soon as you can so I can test.

Then, if you still want to push more code before I merge, let me know.